### ISSUE TITLE
Dotnet Watch Run indefinitely reloads

### ISSUE THREAD

**Issue by jamesikanos** [dotnet/aspnetcore#39073]
### Describe the bug
Calling `dotnet watch run` will detect changes continuously and loop forever.

The first time a file is changed, the Program will rebuild and run. But, after that it will continuously rebuild and reload, even without changing the file.

Workaround, run `dotnet watch run --no-hot-reload` and it will behave as desired (ie: only reloading when a file actually changes).

Environment: WSL2

### To Reproduce
1. `dotnet new console`
2. `dotnet watch run`
3. Save "Program.cs"

Observe that the program will constantly restart.

```
watch : Started
watch : Exited
watch : Waiting for a file to change before restarting dotnet...
watch : Building...
  UnregisteredDeviceSearch -> /home/jamesikanos/scratch/UnregisteredDeviceSearch/bin/Debug/net6.0/UnregisteredDeviceSearch.dll
watch : Started
watch : Exited
watch : Waiting for a file to change before restarting dotnet...
watch : Building...
  UnregisteredDeviceSearch -> /home/jamesikanos/scratch/UnregisteredDeviceSearch/bin/Debug/net6.0/UnregisteredDeviceSearch.dll
watch : Started
watch : Exited
```

### Exceptions (if any)
No exception

### Further technical details

Running on VSCode. This behaviour is also observed when running `dotnet watch run` from Ubuntu terminal.

.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.100/

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  2.1.816 [/usr/share/dotnet/sdk]
  3.1.410 [/usr/share/dotnet/sdk]
  6.0.100 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.28 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.28 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.16 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.28 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.16 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by mrlife** [comment]
I'm seeing the same thing for `dotnet watch`.

```
% dotnet watch
watch : Hot reload enabled. For a list of supported edits, see https://aka.ms/dotnet/hot-reload. Press "Ctrl + R" to restart.
watch : Building...
  Determining projects to restore...
  All projects are up-to-date for restore.
  MyProject -> /Users/username/projects/MyProject/bin/Debug/net6.0/MyProject.dll
watch : Started
watch : Building...
  MyProject -> /Users/username/projects/MyProject/bin/Debug/net6.0/MyProject.dll
watch : Started
watch : Building...
  MyProject -> /Users/username/projects/MyProject/bin/Debug/net6.0/MyProject.dll
watch : Started
watch : Building...
```

**Comment by chris-metz** [comment]
Can confirm the same issue here (Windows 11, dotnet 6.0.101)

**Comment by rgwood** [comment]
Reproduced in a [repl.it](https://repl.it) notebook: https://replit.com/@ripley/DotnetWatchRepro#README.md

This seems like a pretty big bug in `dotnet watch` and I hope it can get addressed.

**Comment by mrlife** [comment]
Just curious, is this an issue for everyone or is there something specific about our projects that are causing it? If the former, I would expect a lot more activity in this issue... unless there are not a lot of people using hot reload.

**Comment by jamesikanos** [comment]
In my experience, this only affects Console projects. Works as expected in Web Projects.

Maybe there aren't many people doing `dotnet watch run` for Console projects?

**Comment by mrlife** [comment]
I am running a Blazor project fwiw.

**Comment by jamesikanos** [comment]
Previous comment rescinded @mrlife . Consider it done.

**Comment by mrlife** [comment]
Could someone comment whether hot reload is supposed to work in a Blazor Server app?

**Comment by CommonLoon102** [comment]
> Could someone comment whether hot reload is supposed to work in a Blazor Server app?

It does work but it takes 45 seconds to complete though.
https://github.com/dotnet/razor-compiler/issues/144

**Comment by Mike-E-angelo** [comment]
45 seconds... if you're lucky @CommonLoon102. :P  I still have yet to see this feature work well or at all in my environment.

**Comment by javiercn** [comment]
@jamesikanos I can't reproduce this behavior with the 6.0.300 SDK. Are you still having issues?

**Comment by javiercn** [comment]
> I'm seeing the same thing for `dotnet watch`.
> 
> ```
> % dotnet watch
> watch : Hot reload enabled. For a list of supported edits, see https://aka.ms/dotnet/hot-reload. Press "Ctrl + R" to restart.
> watch : Building...
>   Determining projects to restore...
>   All projects are up-to-date for restore.
>   MyProject -> /Users/username/projects/MyProject/bin/Debug/net6.0/MyProject.dll
> watch : Started
> watch : Building...
>   MyProject -> /Users/username/projects/MyProject/bin/Debug/net6.0/MyProject.dll
> watch : Started
> watch : Building...
>   MyProject -> /Users/username/projects/MyProject/bin/Debug/net6.0/MyProject.dll
> watch : Started
> watch : Building...
> ```
> 
> Note that I am on macOS 12.0.1 with dotnet 6.0.1.

@mrlife are you still running into this problem with 6.0.300? If so, could you please file a separate issue with concrete details (a repro if possible) so that we can look at it?

**Comment by javiercn** [comment]
> Could someone comment whether hot reload is supposed to work in a Blazor Server app?

Hot reload should work with Blazor Server

**Comment by javiercn** [comment]
> I can reproduce this by adding one line to a fresh Blazor Server project.
> 
> 1. Create a new Blazor Server project (I am on macOS)
> 2. Add the following line to the csproj: `<DefaultAppHostRuntimeIdentifier>win-x64</DefaultAppHostRuntimeIdentifier>`
> 3. Run 'dotnet watch' or 'dotnet watch run' in terminal to see the issue

If you do `dotnet watch --verbose`

You will see that the issue is the process is failing to start:

```
An error occurred trying to start process 'C:\work\scratch\blazorserverwatch\bin\Debug\net6.0\blazorserverwatch' with working directory 'C:\work\scratch\blazorserverwatch'. The specified executable is not a valid application for this OS platform.
```
And that's why there is a loop.

**Comment by javiercn** [comment]
@rgwood In your case the issue with your project is the TFM. It needs to be `net6.0` not `netcoreapp6.0`

**Comment by javiercn** [comment]
In general for folks running into this issue, stop dotnet watch and try to build the app with dotnet build or run it with dotnet run and if that fails, it's likely the issue to start with.

**Comment by javiercn** [comment]
> Reproduced in a [repl.it](https://repl.it) notebook: https://replit.com/@ripley/DotnetWatchRepro#README.md
> 
> This seems like a pretty big bug in `dotnet watch` and I hope it can get addressed.

I correct myself. netcoreapp6.0 works, but the issue is with `<StartupObject>Program</StartupObject>` and top level statements. As far as I can tell this is already addressed in 6.0.300

**Comment by mrlife** [comment]
> @mrlife are you still running into this problem with 6.0.300? If so, could you please file a separate issue with concrete details (a repro if possible) so that we can look at it?

> @mrlife If you do `dotnet watch --verbose`
> 
> You will see that the issue is the process is failing to start:
> 
> ```
> An error occurred trying to start process 'C:\work\scratch\blazorserverwatch\bin\Debug\net6.0\blazorserverwatch' with working directory 'C:\work\scratch\blazorserverwatch'. The specified executable is not a valid application for this OS platform.
> ```
> 
> And that's why there is a loop.

Hi @javiercn, thank you for checking this out. I ran `dotnet watch --verbose` on SDK version 6.0.203, then updated to 6.0.300 and ran the command again. I got the same error each time. I am on macOS and the error seems to indicate it is trying to use a Windows executable.

> watch : Caught top-level exception from hot reload: System.ComponentModel.Win32Exception (8): An error occurred trying to start process '/Users/UserName/Projects/MyProject/bin/Debug/net6.0/MyProject.exe' with working directory '/Users/UserName/MyProject'. Exec format error
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at Microsoft.DotNet.Watcher.Internal.ProcessRunner.RunAsync(ProcessSpec processSpec, CancellationToken cancellationToken)
   at Microsoft.DotNet.Watcher.HotReloadDotNetWatcher.WatchAsync(DotNetWatchContext context, CancellationToken cancellationToken)

**Comment by rgwood** [comment]
Thanks @javiercn. Attempted to confirm the fix but it seems like I'm stuck on v6.0.105 until a newer SDK is published for Linux (on Fedora 36, installing via `dnf install dotnet-sdk-6.0` or `dotnet-install.sh` gets v6.0.105).

**Comment by javiercn** [comment]
@rgwood It'll take some time until the fix is available. I proposed patching it, but it didn't make it into this month patch, so we are looking at July-ish for it to be released.

If you want to verify the fix, you could clone the SDK repo. Run build.sh at the root and then eng\dogfood.sh

That will setup the terminal with a local copy of the dotnet-sdk (including the fix) that contains a dotnet-watch version you can use against your project for validation.

In general, my suggestion would be to try and run `dotnet build` and `dotnet run` to see if those run without issue.
